contracts/SQTGift.sol

```diff
@@ -0,0 +1,147 @@
+
+  RedeemRange[] public redeemRanges;
+  mapping(address => bool) public allowlist;
+  mapping(uint256 => uint256) public sqtRedeemableValue;
+
+  event GiftMinted(address indexed to, uint256 indexed tokenId, string tokenURI, uint256 sqtValue);
+
+  function initialize(uint256 _maxSupply, uint256 _maxSQT, address _sqtoken) external initializer {
+    __Ownable_init();
+    __ERC721_init("SQT Gift", "SQTG");
+    __ERC721URIStorage_init();
+    __ERC721Enumerable_init();
+
+    maxSupply = _maxSupply;
+    maxSQT = _maxSQT;
+    sqtoken = _sqtoken;
+  }
+
+  function setMaxSupply(uint256 _maxSupply) external onlyOwner {
+    maxSupply = _maxSupply;
+  }
+
+  function setMaxSQT(uint256 _maxSQT) external onlyOwner {
+    maxSQT = _maxSQT;
+  }
+
+  function setRedeemer(address _redeemer) external onlyOwner {
+    redeemer = _redeemer;
+  }
+
+  function setDefaultSQTValue(uint256 _defaultSQTValue) external onlyOwner {
+    defaultSQTValue = _defaultSQTValue;
+  }
+
+  function addToAllowlist(address _address) public onlyOwner {
+    allowlist[_address] = true;
+  }
+
+  function removeFromAllowlist(address _address) public onlyOwner {
+    allowlist[_address] = false;
+  }
+
+  function addRedeemRange(uint256 startId, uint256 endId, uint256 sqtValue) public onlyOwner {
+    redeemRanges.push(RedeemRange(startId, endId, sqtValue));
+  }
+
+  function _beforeTokenTransfer(address from, address to, uint256 tokenId, uint256 batchSize) internal override(
+    ERC721Upgradeable, 
+    ERC721EnumerableUpgradeable
+  ) {
+      super._beforeTokenTransfer(from, to, tokenId, batchSize);
+  }
+
+  function supportsInterface(bytes4 interfaceId) public view override(
+    IERC165Upgradeable,
+    ERC721Upgradeable,
+    ERC721EnumerableUpgradeable,
+    ERC721URIStorageUpgradeable
+  ) returns (bool) {
+    return interfaceId == type(ISQTGift).interfaceId || super.supportsInterface(interfaceId);
+  }
+
+  function tokenURI(uint256 tokenId) public view override(
+    ERC721Upgradeable, 
+    ERC721URIStorageUpgradeable
+  ) returns (string memory) {
+      return super.tokenURI(tokenId);
+  }
+
+  function _burn(uint256 tokenId) internal override(ERC721Upgradeable, ERC721URIStorageUpgradeable) {
+      super._burn(tokenId);
+  }
+
+  function _baseURI() internal view virtual override returns (string memory) {
+    return "ipfs://";
+  }
+
+  function _getSQTValueForTokenId(uint256 tokenId) private view returns (uint256) {
+    for (uint i = 0; i < redeemRanges.length; i++) {
+      if (tokenId >= redeemRanges[i].startTokenId && tokenId <= redeemRanges[i].endTokenId) {
+          return redeemRanges[i].sqtValue;
+      }
+    }
+    
+    return defaultSQTValue;
+  }
+
+  function mint(string memory _tokenURI) public {
+    require(allowlist[msg.sender], "Not on allowlist");
+    require(totalSupply() < maxSupply, "Max gift supply reached");
+
+    uint256 tokenId = totalSupply() + 1;
+    uint256 sqtValue = _getSQTValueForTokenId(tokenId);
```

Comment:
this sqtValue should be written as an attributes, then we can use a standard way to read it, instead of  `getSQTRedeemableValue()`